src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java

```diff
@@ -207,31 +226,52 @@ public void close(int code) throws IOException {
 
     @Override
     public void close(int code, String reason) throws IOException {
-        byte[] reasonBytes = null;
+        if (outputState == CLOSED) {
+            return;
+        }
+
+        try {
+            stateLock.lock();
 
-        if (code != 0) {
-            // Verify code and reason against RFC 6455.
-            // If code is present, it must equal to 1000 or in range 3000 to 4999
-            if (code != 1000 && (code < 3000 || code > 4999)) {
-                throw new IOException(MSG_INVALID_CLOSE_CODE);
+            if (outputState == CLOSED) {
+                return;
             }
 
-            if (reason != null && reason.length() > 0) {
-                reasonBytes = reason.getBytes(UTF_8);
+            byte[] reasonBytes = null;
+
+            if (code != 0) {
+                // Verify code and reason against RFC 6455.
+                // If code is present, it must equal to 1000 or in range 3000 to 4999
+                if (code != 1000 && (code < 3000 || code > 4999)) {
```

Comment:
Note that the aforementioned check is for the code coming in from the application.